File: Queue.hs
import Control.DeepSeq (NFData, rnf, deepseq)
import System.Random (randomRIO)
import Control.Monad (replicateM)
import Data.Time.Clock (getCurrentTime, diffUTCTime)
import System.Mem (performGC)
import GHC.Stats (getRTSStats, max_mem_in_use_bytes)

data Queue a = Queue [a] [a] deriving (Show)

instance NFData a => NFData (Queue a) where
    rnf (Queue front rear) = rnf front seq rnf rear

emptyQueue :: Queue a
emptyQueue = Queue [] []

enqueue :: a -> Queue a -> Queue a
enqueue x (Queue front rear) = Queue front (x:rear)

dequeue :: Queue a -> (a, Queue a)
dequeue (Queue [] []) = error "Empty queue"
dequeue (Queue [] rear) = dequeue (Queue (reverse rear) [])
dequeue (Queue (x:xs) rear) = (x, Queue xs rear)

randomList :: Int -> IO [Int]
randomList n = replicateM n $ randomRIO (1, 1000000)

estimateMemory :: Int -> Int
estimateMemory elements = elements * 8  

benchmarkQueueOperations :: Int -> IO ()
benchmarkQueueOperations n = do
    values <- randomList n
    start <- getCurrentTime
    performGC
    let queue = foldl (flip enqueue) emptyQueue values
    queue deepseq return ()
    end <- getCurrentTime
    performGC
    stats <- getRTSStats
    let duration = diffUTCTime end start
        memoryUsage = max_mem_in_use_bytes stats
        estimatedMemory = estimateMemory n
    putStrLn $ "Benchmark for " ++ show n ++ " elements:"
    putStrLn $ "Time taken: " ++ show duration ++ " seconds"
    putStrLn $ "Memory: " ++ show memoryUsage ++ " bytes"
    putStrLn $ "Estimated memory usage: " ++ show estimatedMemory ++ " bytes"
    putStrLn "--------------------------------------------------"

main :: IO ()
main = do
    benchmarkQueueOperations 100000
    benchmarkQueueOperations 200000
    benchmarkQueueOperations 500000

File: Queue.cpp
#include <iostream>
#include <memory>
#include <functional>
#include <stdexcept>
#include <chrono>
#include <vector>
#include <algorithm>

template<typename T>
class Queue {
public:
    struct Node {
        T value;
        std::shared_ptr<Node> next;
        Node(T val, std::shared_ptr<Node> nxt = nullptr) : value(val), next(nxt) {}
    };
    std::shared_ptr<Node> front;
    std::shared_ptr<Node> back;

public:
    Queue() : front(nullptr), back(nullptr) {}

    Queue enqueue(T value) const {
        auto newNode = std::make_shared<Node>(value);
        if (back) {
            back->next = newNode;
        }
        return Queue(front ? front : newNode, newNode);
    }

    Queue dequeue() const {
        if (!front) throw std::out_of_range("Empty queue");
        return Queue(front->next, front->next ? back : nullptr);
    }

    T peek() const {
        if (!front) throw std::out_of_range("Empty queue");
        return front->value;
    }

    bool isEmpty() const {
        return front == nullptr;
    }

private:
    Queue(std::shared_ptr<Node> f, std::shared_ptr<Node> b) : front(f), back(b) {}
};

void benchmarkQueue(size_t n) {
    Queue<int> queue;
    std::vector<int> values(n);
    std::generate(values.begin(), values.end(), []() { return std::rand() % 1000000; });

    auto start = std::chrono::high_resolution_clock::now();

   
    for (int value : values) {
        queue = queue.enqueue(value);
    }

    size_t memoryUsage = n * (sizeof(int) + sizeof(std::shared_ptr<typename Queue<int>::Node>));

    while (!queue.isEmpty()) {
        queue = queue.dequeue();
    }

    auto stop = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(stop - start);

    std::cout << "Time taken to enqueue and dequeue " << n << " elements: " << duration.count() << " milliseconds\n";
    std::cout << "Memory used: " << memoryUsage << " bytes\n";
}

int main() {
    benchmarkQueue(100000);
    benchmarkQueue(200000);
    benchmarkQueue(500000);
    return 0;
}
